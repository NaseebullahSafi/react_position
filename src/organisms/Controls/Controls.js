import React, { useState, useEffect } from 'react'
import plus from '../../assets/icons/plus.svg'
import './Controls.scss'
import Rover from '../../molecules/Rover/Rover'

/**
 * Will render the control container which will allow users to
 * interact with n number of rovers
 * @param {Object} 
 */
const Controls = ({ dimension, onUpdate }) => {
    const defaultRover = {
        x: 0,
        y: 0,
        face: 'N',
        path: ''
    }
    const [rovers, modifyRover] = useState([defaultRover]);

    // call this when adding new rover
    const addRover = rover => modifyRover([...rovers, rover])

    // call this when updating an existing rover
    const updateRover = updateRovers => modifyRover(updateRovers)

    // call this when deleting the rover
    const deleteRover = i => modifyRover(rovers.filter((_, j) => 1 !== j))

    // let us update on mounting the component
    useEffect(() => onUpdate(rovers))

    return (
        <div className="controls">
            <h1 className="controls__title">Controls</h1>

            {rovers.map((rover, i) => (
                <Rover
                    key={i}
                    mars={dimension}
                    {...rover}
                    onChange={newPos => {
                        updateRover(rovers.map((r, j) => {
                            if (i === j) r = newPos
                            return r
                        }))
                    }}
                    onDelete={() => deleteRover(i)}
                />
            ))}

            <div className="controls__add" onClick={e => {
                e.preventDefault()
                addRover(defaultRover)
            }}>
                <img src={plus} />
            </div>
        </div>
    )
}

export default Controls;