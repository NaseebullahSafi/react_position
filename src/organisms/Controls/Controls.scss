.controls {
  align-items: center;
  align-self: stretch;
  background-color: #4d5961;
  border-radius: 5px;
  box-shadow: 0 0 10px -3px #48545a;
  display: flex;
  flex: 0 0 30%;
  flex-direction: column;
  justify-content: flex-start;
  padding: 10px;
  position: relative;

  &__title {
    color: white;
    letter-spacing: 10px;
  }

  &__add {
    align-items: center;
    background-color: #d06262;
    border-radius: 5px;
    box-shadow: 0 0 10px -2px #4d4f56;
    display: flex;
    height: 45px;
    justify-content: center;
    margin: 20px 0;
    transition: all 0.15s ease-in-out;
    width: 71%;

    &:hover {
      cursor: pointer;
      transform: scale(1.05);
    }
  }
}
