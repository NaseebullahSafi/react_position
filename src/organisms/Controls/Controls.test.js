import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Controls from './Controls';

afterEach(cleanup)

describe('Controls Component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(<Controls onUpdate={() => {}} />);
    expect(asFragment()).toMatchSnapshot()
  })
})