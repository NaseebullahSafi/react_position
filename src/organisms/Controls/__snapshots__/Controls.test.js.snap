// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls Component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="controls"
  >
    <h1
      class="controls__title"
    >
      Controls
    </h1>
    <div
      class="rover"
    >
      <div
        class="rover__icon rover__icon--close"
      >
        <svg
          fill="none"
          height="24"
          stroke="#4d5961"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </div>
      <div
        class="rover__pos-container"
      >
        <div>
          <span>
            Pos ( x, y, face ): 
          </span>
          <input
            class="input"
            placeholder="0 0 N"
          />
        </div>
        <div
          class="rover__cta-container"
        >
          <div
            class="rover__cta-item"
          >
            <svg
              fill="none"
              height="24"
              stroke="#4d5961"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="19"
                x2="5"
                y1="12"
                y2="12"
              />
              <polyline
                points="12 19 5 12 12 5"
              />
            </svg>
          </div>
          <div
            class="rover__cta-item"
          >
            <svg
              fill="none"
              height="24"
              stroke="#d06262"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="2"
              />
              <path
                d="M16.24 7.76a6 6 0 0 1 0 8.49m-8.48-.01a6 6 0 0 1 0-8.49m11.31-2.82a10 10 0 0 1 0 14.14m-14.14 0a10 10 0 0 1 0-14.14"
              />
            </svg>
          </div>
          <div
            class="rover__cta-item"
          >
            <svg
              fill="none"
              height="24"
              stroke="#4d5961"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
              <polyline
                points="12 5 19 12 12 19"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="rover__path"
      >
        <input
          class="input"
          placeholder="Command path"
          readonly=""
          value=""
        />
      </div>
    </div>
    <div
      class="controls__add"
    >
      <img
        src="plus.svg"
      />
    </div>
  </div>
</DocumentFragment>
`;
